plugin/net/cni.go

```diff
@@ -152,7 +152,7 @@ func (c *CNIPlugin) CmdAdd(args *skel.CmdArgs) error {
 
 func setupRoutes(link netlink.Link, name string, ipnet net.IPNet, gw net.IP, routes []*types.Route) error {
 	var err error
-	if routes == nil { // If config says nothing about routes, add a default one
+	if routes == nil || len(routes) == 0 { // If config says nothing about routes, add a default one
```

Comment:
This change makes the comment wrong - if it says _nothing_ then the value will be nil.

This is a change to the intended behaviour - can you check where the empty list is coming from, so we can understand why we need to change?

It would also be good to have some way to request the default Weave Net route to be added even when an IPAM plugin has added some routes.

Reply:
>  can you check where the empty list is coming from, so we can understand why we need to change?
Weave IPAM plug-in is just copying the routes from configuration to response.
https://github.com/weaveworks/weave/blob/master/plugin/ipam/cni.go#L60

Same is case with `host-local` plugin https://github.com/containernetworking/plugins/blob/master/plugins/ipam/host-local/main.go#L114

> This is a change to the intended behaviour 

In which case I will move the check in to Weave's IPAM so _it does not return empty list_ when _config says nothing about routes_

Reply:
But where was the zero-length slice created?  It's not in the config in #3331

Reply:
Since the CNI go [definition](https://github.com/containernetworking/cni/blob/master/pkg/types/current/types.go#L140) for IPAM response has Routes as array, json un-marshalling is creating empty slice

```go
// Result is what gets returned from the plugin (via stdout) to the caller
type Result struct {
	CNIVersion string         `json:"cniVersion,omitempty"`
	Interfaces []*Interface   `json:"interfaces,omitempty"`
	IPs        []*IPConfig    `json:"ips,omitempty"`
	Routes     []*types.Route `json:"routes,omitempty"`
	DNS        types.DNS      `json:"dns,omitempty"`
}
```

Reply:
Zero-length slice if getting created in libcni, Please see the comment 

https://github.com/weaveworks/weave/issues/3331#issuecomment-408388766

I will close this PR as behaviour in weave CNI and IPAM plug-in's are intended.